internal/validators/res/resource_validator.go

```diff
@@ -0,0 +1,157 @@
+// Package res contains the logic for validating resource rules
+package res
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/canonical/gomaasclient/api"
+	"github.com/canonical/gomaasclient/entity"
+	"github.com/go-logr/logr"
+
+	"github.com/validator-labs/validator-plugin-maas/api/v1alpha1"
+	"github.com/validator-labs/validator-plugin-maas/internal/constants"
+	"github.com/validator-labs/validator-plugin-maas/internal/utils"
+	"github.com/validator-labs/validator/pkg/types"
+)
+
+type resource struct {
+	NumCPU int
+	RAM    int
+	Disk   int
+	Pool   string
+	Tags   []string
+}
+
+// ResourceRulesService is a service for reconciling resource rules
+type ResourceRulesService struct {
+	log logr.Logger
+	api api.Machines
+}
+
+// NewResourceRulesService returns a ResourceRulesService
+func NewResourceRulesService(log logr.Logger, api api.Machines) *ResourceRulesService {
+	return &ResourceRulesService{
+		log: log,
+		api: api,
+	}
+}
+
+// ReconcileMaasInstanceResourceRule reconciles a MAAS instance resource rule from the MaasValidator config
+func (s *ResourceRulesService) ReconcileMaasInstanceResourceRule(rule v1alpha1.ResourceAvailabilityRule, seen []string) (*types.ValidationRuleResult, error) {
+	errs := make([]error, 0)
+	details := make([]string, 0)
+
+	vr := utils.BuildValidationResult(rule.Name, constants.ValidationTypeResource)
+
+	// do not process an AZ more than once
+	if containsString(seen, rule.AZ) {
+		errs = append(errs, fmt.Errorf("availability zone %s already validated", rule.AZ))
+	} else {
+		errs, details = s.findMachineResources(rule)
+	}
+
+	utils.UpdateResult(vr, errs, constants.ErrImageNotFound, details...)
+
+	if len(errs) > 0 {
+		return vr, errs[0]
+	}
+	return vr, nil
+}
+
+// contains checks if a slice contains a string
+func containsString(s []string, e string) bool {
```

Comment:
You can use `slices.Contains`, but it would be even better (more efficient) to just use a map instead.